### ISSUE TITLE
[regression] JPEGd 422V (440P) wrong Cr output

### ISSUE THREAD

**Issue by uartie** [intel/media-driver#1303]
## System information
- Impacted GPU platform:

  KBL, CFL, DG1

- Host machine if it's descreted card:
- **GOOD** libva/libva-utils/gmmlib/media-driver version?

  intel-media-driver: e2011ffcb5dc2d04ffd320ebcfe57fb55522baa7

  cartwheel ffmpeg: https://github.com/intel-media-ci/cartwheel-ffmpeg/commit/3ef607590e782c9266308a66b73979c740558910

- **BAD** libva/libva-utils/gmmlib/media-driver version?

  intel-media-driver: 497d9864aaba83214cbc2e6ae8cee4151934fe43

### Describe the current behavior

```shell
ffmpeg -hwaccel vaapi -init_hw_device vaapi=hw:/dev/dri/renderD128 -hwaccel_output_format yuv440p \
  -hwaccel_flags allow_profile_mismatch -v verbose -i smpte_422V.jpeg -pix_fmt yuv440p -f rawvideo \
  -vsync passthrough -autoscale 0 -vframes 1 -y smpte_422V_720x480_422V.yuv
```

- Output on KBL, CFL, DG1 has corrupt Cr output:

  ![output-KBL](https://user-images.githubusercontent.com/677859/144091755-c1db1b29-c2d8-4c5b-9b13-6e19538e68f7.jpeg)

- Output on TGL is good:

  ![output-TGL](https://user-images.githubusercontent.com/677859/144091869-2268a92e-db3c-4a18-84f1-ca22f2ba121d.jpeg)

### Describe the expected behavior

- All platforms output should look like:

  ![output-TGL](https://user-images.githubusercontent.com/677859/144091869-2268a92e-db3c-4a18-84f1-ca22f2ba121d.jpeg)

**Comment by uartie** [comment]
Here is the 422V input file:

![smpte_422V](https://user-images.githubusercontent.com/677859/144092383-39bd8631-d907-4566-8073-f5a0407e038a.jpeg)

**Comment by XinfengZhang** [comment]
seems it is because the incorrect UV offset.   which function is used to read final output? vaGetImage , or vaDeriveImage?

**Comment by uartie** [comment]
> seems it is because the incorrect UV offset. which function is used to read final output? vaGetImage , or vaDeriveImage?

@xhaihao do you have answer?

**Comment by xhaihao** [comment]
@uartie @XinfengZhang  FFmpeg uses vaGetImage() to retrieve data from gfx memory.

**Comment by xhaihao** [comment]
@cqian2 @XinfengZhang any update for this issue ?

**Comment by cqian2** [comment]
@xhaihao , we are still syncing it internally. Will update here after it is confirmed or we have conclusion for it. Thx.

**Comment by fenhu** [comment]
Would you try belwo patch?

intel@intel-OptiPlex-9020-0:~/gfx/opensource/media-driver$ git diff ./
diff --git a/media_driver/linux/common/ddi/media_libva.cpp b/media_driver/linux/common/ddi/cpp
index 865f9a97..99975efe 100755
--- a/media_driver/linux/common/ddi/media_libva.cpp
+++ b/media_driver/linux/common/ddi/media_libva.cpp
@@ -5174,7 +5174,7 @@ static VAStatus DdiMedia_CopySurfaceToImage(
     PDDI_MEDIA_CONTEXT mediaCtx = DdiMedia_GetMediaContext(ctx);
     DDI_CHK_NULL(mediaCtx,  "nullptr mediaCtx.",    VA_STATUS_ERROR_INVALID_CONTEXT);
     DDI_CHK_NULL(surface,  "nullptr meida surface.", VA_STATUS_ERROR_INVALID_BUFFER);
-
+    uint32_t flag = MOS_LOCKFLAG_READONLY;
     VAStatus vaStatus = VA_STATUS_SUCCESS;
     //Lock Surface
     if ((Media_Format_CPU != surface->format))
@@ -5186,7 +5186,11 @@ static VAStatus DdiMedia_CopySurfaceToImage(
             DDI_NORMALMESSAGE("surface Decompression fail, continue next steps.");
         }
     }
-    void *surfData = DdiMediaUtil_LockSurface(surface, MOS_LOCKFLAG_READONLY);
+
+    if (image->format.fourcc != VA_FOURCC_NV12)
+       flag = flag | MOS_LOCKFLAG_NO_SWIZZLE;
+
+    void *surfData = DdiMediaUtil_LockSurface(surface, flag);
     if (surfData == nullptr)
     {
         DDI_ASSERTMESSAGE("nullptr surfData.");
@@ -5205,8 +5209,18 @@ static VAStatus DdiMedia_CopySurfaceToImage(
     uint8_t *ySrc = nullptr;
     uint8_t *yDst = (uint8_t*)imageData;

-    ySrc = (uint8_t*)surfData;
+    uint8_t *swizzleData = (uint8_t*)MOS_AllocMemory(surface->data_size);

+    if (!surface->pMediaCtx->bIsAtomSOC && surface->TileType != I915_TILING_NONE && image-cc != VA_FOURCC_NV12)
+    {
+        SwizzleSurface(surface->pMediaCtx, surface->pGmmResourceInfo, surfData, (MOS_TILE_->TileType, (uint8_t *)swizzleData, false);
+        ySrc = swizzleData;
+    }
+    else
+    {
+        ySrc = (uint8_t*)surfData;
+    }
+
     DdiMedia_CopyPlane(yDst, image->pitches[0], ySrc, surface->iPitch, image->height);
     if (image->num_planes > 1)
     {
@@ -5228,6 +5242,7 @@ static VAStatus DdiMedia_CopySurfaceToImage(
         }
     }

+    MOS_FreeMemory(swizzleData);
     vaStatus = DdiMedia_UnmapBuffer(ctx, image->buf);
     if (vaStatus != VA_STATUS_SUCCESS)
     {
@@ -5303,7 +5318,10 @@ VAStatus DdiMedia_GetImage(
     VASurfaceID output_surface = surface;

     if (inputSurface->format != DdiMedia_OsFormatToMediaFormat(vaimg->format.fourcc, vaimgha_mask) ||
-        width != vaimg->width || height != vaimg->height)
+        width != vaimg->width || height != vaimg->height &&
+        (vaimg->format.fourcc != VA_FOURCC_444P &&
+        vaimg->format.fourcc != VA_FOURCC_422V &&
+        vaimg->format.fourcc != VA_FOURCC_422H))
     {
         VAContextID context = VA_INVALID_ID;
         //Create VP Context.

**Comment by fenhu** [comment]
There is the PR:
https://github.com/intel-innersource/drivers.gpu.unified/pull/52661
verified both on KBL and DG1.   JPEGd works.

**Comment by XinfengZhang** [comment]
it should be fixed by fce936af16da67ccb9327ad1fa023cec769df00a

**Comment by uartie** [comment]
Thanks. Fix confirmed.
